File: components/agentic-commerce-worker/src/http.rs
/// Synchronous HTTP GET and POST using raw WASI poll primitives.
///
/// `fut.subscribe().block()` is the WASI equivalent of block_on — it
/// synchronously polls the future until the response arrives, using the
/// same WASI poll mechanism that wstd's block_on uses internally.
///
/// We implement this directly instead of via wstd because wstd's async
/// runtime pulls in `wasi:random@0.2.9` which the WAVS operator world
/// does not provide.
use crate::bindings::wasi::http::outgoing_handler;
use crate::bindings::wasi::http::types::{
    Fields, IncomingBody, Method, OutgoingBody, OutgoingRequest, Scheme,
};
use crate::bindings::wasi::io::streams::StreamError;

pub fn get(url: &str) -> Result<Vec<u8>, String> {
    let (scheme, authority, path_query) = parse_url(url)?;
    let headers = Fields::new();
    let req = OutgoingRequest::new(headers);
    req.set_method(&Method::Get).map_err(|_| "set method")?;
    req.set_scheme(Some(&scheme)).map_err(|_| "set scheme")?;
    req.set_authority(Some(&authority)).map_err(|_| "set authority")?;
    req.set_path_with_query(Some(&path_query)).map_err(|_| "set path")?;
    send_request(req)
}

pub fn post_json(url: &str, body_bytes: &[u8]) -> Result<Vec<u8>, String> {
    let (scheme, authority, path_query) = parse_url(url)?;
    let headers = Fields::new();
    headers.append(&"content-type".to_string(), &b"application/json".to_vec())
        .map_err(|e| format!("set content-type: {:?}", e))?;
    headers.append(&"content-length".to_string(), &body_bytes.len().to_string().into_bytes())
        .map_err(|e| format!("set content-length: {:?}", e))?;

    let req = OutgoingRequest::new(headers);
    req.set_method(&Method::Post).map_err(|_| "set method POST")?;
    req.set_scheme(Some(&scheme)).map_err(|_| "set scheme")?;
    req.set_authority(Some(&authority)).map_err(|_| "set authority")?;
    req.set_path_with_query(Some(&path_query)).map_err(|_| "set path")?;

    {
        let outgoing_body = req.body().map_err(|_| "get outgoing body")?;
        let stream = outgoing_body.write().map_err(|_| "get write stream")?;
        stream.blocking_write_and_flush(body_bytes)
            .map_err(|e| format!("write body: {:?}", e))?;
        drop(stream);
        OutgoingBody::finish(outgoing_body, None)
            .map_err(|e| format!("finish body: {:?}", e))?;
    }

    send_request(req)
}

/// Publish plain-text content to paste.rs and return the paste URL.
/// Uses PUT https://paste.rs/ with raw text body — returns a plain-text URL.
pub fn publish_paste(content: &str) -> Result<String, String> {
    let body_bytes = content.as_bytes();
    let (scheme, authority, path_query) = parse_url("https://paste.rs/")?;
    let headers = Fields::new();
    headers.append(&"content-type".to_string(), &b"text/plain".to_vec())
        .map_err(|e| format!("set content-type: {:?}", e))?;
    headers.append(&"content-length".to_string(), &body_bytes.len().to_string().into_bytes())
        .map_err(|e| format!("set content-length: {:?}", e))?;

    let req = OutgoingRequest::new(headers);
    req.set_method(&Method::Put).map_err(|_| "set method PUT")?;
    req.set_scheme(Some(&scheme)).map_err(|_| "set scheme")?;
    req.set_authority(Some(&authority)).map_err(|_| "set authority")?;
    req.set_path_with_query(Some(&path_query)).map_err(|_| "set path")?;

    {
        let outgoing_body = req.body().map_err(|_| "get outgoing body")?;
        let stream = outgoing_body.write().map_err(|_| "get write stream")?;
        stream.blocking_write_and_flush(body_bytes)
            .map_err(|e| format!("write body: {:?}", e))?;
        drop(stream);
        OutgoingBody::finish(outgoing_body, None)
            .map_err(|e| format!("finish body: {:?}", e))?;
    }

    let response_bytes = send_request(req)?;
    let url = String::from_utf8(response_bytes)
        .map_err(|e| format!("paste.rs response not UTF-8: {}", e))?
        .trim()
        .to_string();

    if url.is_empty() || !url.starts_with("http") {
        return Err(format!("paste.rs returned unexpected response: {:?}", url));
    }

    Ok(url)
}

fn send_request(req: OutgoingRequest) -> Result<Vec<u8>, String> {
    let fut = outgoing_handler::handle(req, None)
        .map_err(|e| format!("HTTP request failed: {:?}", e))?;

    // Block synchronously using WASI poll — equivalent to block_on
    fut.subscribe().block();

    let resp = fut.get()
        .ok_or("future not ready after block")?
        .map_err(|_| "response future error")?
        .map_err(|e| format!("HTTP error: {:?}", e))?;

    let status = resp.status();
    if status < 200 || status >= 300 {
        return Err(format!("HTTP {}", status));
    }

    let body = resp.consume().map_err(|_| "consume body")?;
    let stream = body.stream().map_err(|_| "body stream")?;

    let mut bytes = Vec::new();
    let mut empty_reads = 0;
    loop {
        stream.subscribe().block();
        match stream.read(65536) {
            Ok(chunk) if chunk.is_empty() => {
                empty_reads += 1;
                if empty_reads > 32 { break; }
            }
            Ok(chunk) => { bytes.extend_from_slice(&chunk); empty_reads = 0; }
            Err(StreamError::Closed) => break,
            Err(StreamError::LastOperationFailed(e)) => {
                return Err(format!("read error: {:?}", e));
            }
        }
    }
    drop(stream);
    let _trailers = IncomingBody::finish(body);
    Ok(bytes)
}

fn parse_url(url: &str) -> Result<(Scheme, String, String), String> {
    let (scheme, rest) = if let Some(s) = url.strip_prefix("https://") {
        (Scheme::Https, s)
    } else if let Some(s) = url.strip_prefix("http://") {
        (Scheme::Http, s)
    } else {
        return Err(format!("unsupported URL scheme: {}", url));
    };
    let (authority, path_query) = match rest.find('/') {
        Some(pos) => (rest[..pos].to_string(), rest[pos..].to_string()),
        None => (rest.to_string(), "/".to_string()),
    };
    Ok((scheme, authority, path_query))
}

File: components/golang-evm-price-oracle/src/http.go
// Package main — raw WASI HTTP client.
//
// Uses the generated wasi:http/outgoing-handler@0.2.0 bindings directly,
// avoiding any third-party HTTP library (which has cm version conflicts with TinyGo).
// This mirrors the approach used in the Rust WAVS components.
package main

import (
	"fmt"
	"strings"

	httphandler "github.com/Lay3rLabs/wavs-examples/components/golang-evm-price-oracle/gen/wasi/http/outgoing-handler"
	httptypes "github.com/Lay3rLabs/wavs-examples/components/golang-evm-price-oracle/gen/wasi/http/types"
	"github.com/Lay3rLabs/wavs-examples/components/golang-evm-price-oracle/gen/wasi/io/poll"
	"go.bytecodealliance.org/cm"
)

// httpGet performs a synchronous HTTP GET and returns the response body.
func httpGet(url string) ([]byte, error) {
	scheme, authority, pathQuery, err := parseURL(url)
	if err != nil {
		return nil, err
	}

	headers := httptypes.NewFields()
	req := httptypes.NewOutgoingRequest(headers)

	req.SetMethod(httptypes.MethodGet())
	req.SetScheme(cm.Some(scheme))
	req.SetAuthority(cm.Some(authority))
	req.SetPathWithQuery(cm.Some(pathQuery))

	// Set Accept header
	acceptKey := httptypes.FieldKey("Accept")
	acceptVal := httptypes.FieldValue(cm.ToList([]byte("application/json")))
	headers.Append(acceptKey, acceptVal)

	result := httphandler.Handle(req, cm.None[httptypes.RequestOptions]())
	if result.IsErr() {
		return nil, fmt.Errorf("HTTP handle: %v", result.Err())
	}

	future := result.OK()
	// wasi:io/poll@0.2.9 replaced pollable.Block() with poll.Poll()
	sub := future.Subscribe()
	poll.Poll(cm.ToList([]poll.Pollable{sub}))
	sub.ResourceDrop()

	getResult := future.Get()
	if getResult.None() {
		return nil, fmt.Errorf("future not ready after block")
	}

	inner := getResult.Some()
	if inner.IsErr() {
		return nil, fmt.Errorf("future error: %v", inner.Err())
	}

	resp := inner.OK()
	if resp.IsErr() {
		return nil, fmt.Errorf("HTTP response error: %v", resp.Err())
	}

	response := resp.OK()
	status := response.Status()
	if status < 200 || status >= 300 {
		return nil, fmt.Errorf("HTTP %d from %s", status, url)
	}

	bodyResult := response.Consume()
	if bodyResult.IsErr() {
		return nil, fmt.Errorf("consume body: %v", bodyResult.Err())
	}
	body := bodyResult.OK()

	streamResult := body.Stream()
	if streamResult.IsErr() {
		return nil, fmt.Errorf("body stream: %v", streamResult.Err())
	}
	stream := streamResult.OK()

	var buf []byte
	emptyReads := 0
	for {
		// wasi:io/poll@0.2.9: use Poll() instead of pollable.Block()
		sub := stream.Subscribe()
		poll.Poll(cm.ToList([]poll.Pollable{sub}))
		sub.ResourceDrop()
		readResult := stream.Read(65536)
		if readResult.IsErr() {
			// StreamErrorClosed = normal end of stream
			break
		}
		chunk := readResult.OK()
		if len(chunk.Slice()) == 0 {
			emptyReads++
			if emptyReads > 32 {
				break
			}
			continue
		}
		buf = append(buf, chunk.Slice()...)
		emptyReads = 0
	}

	// wasi:http@0.2.9: IncomingBody.finish() removed; just drop the resource
	body.ResourceDrop()
	return buf, nil
}

// parseURL splits a URL into (scheme, authority, pathWithQuery).
func parseURL(url string) (httptypes.Scheme, string, string, error) {
	var scheme httptypes.Scheme
	var rest string

	if after, ok := strings.CutPrefix(url, "https://"); ok {
		scheme = httptypes.SchemeHTTPS()
		rest = after
	} else if after, ok := strings.CutPrefix(url, "http://"); ok {
		scheme = httptypes.SchemeHTTP()
		rest = after
	} else {
		return httptypes.Scheme{}, "", "", fmt.Errorf("unsupported URL scheme: %s", url)
	}

	var authority, pathQuery string
	if idx := strings.IndexByte(rest, '/'); idx >= 0 {
		authority = rest[:idx]
		pathQuery = rest[idx:]
	} else {
		authority = rest
		pathQuery = "/"
	}

	return scheme, authority, pathQuery, nil
}
